### ISSUE TITLE
Support for ParamArray After Optional Parameters

### ISSUE THREAD

**Issue by jpbro** [WaynePhillipsEA/twinbasic#649]
**Is your feature request related to a problem? Please describe.**
In VB6, a procedure declaration Like this is illegal:

```
Public Sub StartDoc(Optional ByVal DocType As String = "html", _
                    Optional ByVal Language As String = "en-US", _
                    Optional ByVal Charset As String = "utf-8", _
                    Optional ByVal Viewport As String = "width=device-width, initial-scale=1.0", _
                    ParamArray HeadTagAttributes() As Variant)
```

I may be missing something obvious, but I don't see a reason why it shouldn't be allowed. For example, calling the above method like:

` StartDoc , , , , "id", "pagehead" `

It would use the defaults for all the optional parameters, and then stuff "id" and "pagehead" into the HeadTagAttributes variant parameter array.

So am I being a dope, or should such a construction be allowed in tB?

**Describe the solution you'd like**
I would like to be able to use ParamArray after Optional parameters in a procedure declaration if it is possible, and I'm not missing a good reason why it can't be done.

**Describe alternatives you've considered**
Alternative would be to not implement this feature because I've missed something glaringly obvious, or it's just an awful idea and I should probably turn in my keyboard and programmer's badge.

**Comment by retailcoder** [comment]
I feel like this is a bad idea, but can't quite put the finger on _why_... I would try to leverage method overloading, and/or receive an actual array (optional?) instead of a paramarray.

Optional parameters are often an indication of a method doing too many things, or of a missing data type or structure. Perhaps that method wants some ElementInfo parameter instead?

That said IIRC this limitation if we can call it that, is inherent to COM, so allowing paramarray to be specified after an optional parameter would mean the method couldn't export to COM clients.

**Comment by bclothier** [comment]
IMO, this is probably better solved by giving UDT their own constructors. See #600 . tB would also be well-served by having native support for associative arrays, which is very common thing in web application (whether backend or frontend). I think we just fake it with `Scripting.Dictionary` but would be nice to be able to get a default prototype (hence the link to the other issue).

**Comment by jpbro** [comment]
I'll start at the end:

> That said IIRC this limitation if we can call it that, is inherent to COM, so allowing paramarray to be specified after an optional parameter would mean the method couldn't export to COM clients.

If that's the case, then the idea is a total non-starter for sure.

> I feel like this is a bad idea, but can't quite put the finger on _why_... I would try to leverage method overloading, and/or receive an actual array (optional?) instead of a paramarray.

I'm not a fan of it either (at least for 99.9% of cases), but for the thing I'm working on it felt kinda right. I'm putting together a few classes that will make it easier to generate HTML in VB/tB, and when I'm writing HTML I think in a kind of linear stream. If was hoping to do studd like this:

```
With .AddTag("p", "id", "someid", "class", "first,important")
   .Content = "Hello there."
End With ' Automatically closes the tag
```

As opposed to:

```
Dim Attribs(3) As String
 
Attribs(0) = "id"
Attribs(1) = "someid"
Attribs(2) = "class"
Attribs(3) = "first,important"

With .AddTag("p", Attribs)
   .Content = "Hello there."
End With
```

But I suppose I could just go with:

```
With .AddTag("p", Array("id", "someid", "class", "first,important"))
   .Content = "Hello there."
End With ' Automatically closes the tag
```
Anyway, I'm still in the daydreaming stage, so I'll figure out a better (or perhaps crazier) way :)

> Optional parameters are often an indication of a method doing too many things, or of a missing data type or structure. Perhaps that method wants some ElementInfo parameter instead?

Yeah, I agree in the vast majority of cases. I'm just trying to make something that makes it quick and easy to bang out HTML in a stream of consciousness kind of way. The aim is speed over beauty for things like prototyping ideas, making minimum viable products quickly, etc.... I just thought, hey we have optionals and we have paramarrays - why not both?

**Comment by jpbro** [comment]
> EDIT: opened #650 to discuss the idea of associative array which I think would be much better solution than a bunch of optionals & paramarray.

#650 looks like a really cool idea IMO.

**Comment by EduardoVB** [comment]
> Optional parameters are often an indication of a method doing too many things, or of a missing data type or structure.

Sometimes yes, that's the case.
But I don't think that per se it means a bad design. 
Lot of VBx power is because it does not require to specify everything, but still providing ways to customize.

Some Examples:

```
Sub PaintPicture(ByVal Picture As IPictureDisp, ByVal X1 As Single, ByVal Y1 As Single, Optional ByVal Width1 As Variant, Optional ByVal Height1 As Variant, Optional ByVal X2 As Variant, Optional ByVal Y2 As Variant, Optional ByVal Width2 As Variant, Optional ByVal Height2 As Variant, Optional ByVal Opcode As Variant)

Function MsgBox(Prompt As Variant, Optional Buttons As VbMsgBoxStyle = vbOKOnly, Optional Title As Variant, Optional HelpFile As Variant, Optional Context As Variant) As VbMsgBoxResult
```
If parameters are not provided, it uses defaults.

**Comment by WaynePhillipsEA** [comment]
Nice idea, but as mentioned, this is a COM restriction.  See the FUNCFLAGS structure which is used in type libraries (and thus everything in COM is based around): https://docs.microsoft.com/en-us/windows/win32/api/oaidl/ns-oaidl-funcdesc   

In particular this note, which details the behaviour of the cParamsOpt count field:

- The cParamsOpt field specifies the form of optional parameters accepted by the function, as follows:

- A value of 0 specifies that no optional arguments are supported.
- A value of –1 specifies that the method's last parameter is a pointer to a safe array of variants. Any number of variant arguments greater than cParams –1 must be packaged by the caller into a safe array and passed as the final parameter. This array of optional parameters must be freed by the caller after control is returned from the call.
Any other number indicates that the last n parameters of the function are variants and do not need to be specified by the caller explicitly. The parameters left unspecified should be filled in by the compiler or interpreter as variants of type VT_ERROR with the value DISP_E_PARAMNOTFOUND.

In short; that field is used to indicate a ParamArray (-1), **_or_** a specific number of optional arguments (mutually exclusive).

**Comment by jpbro** [comment]
Thanks for the detailed response @WaynePhillipsEA !
